### PR TITLE
Added `scan(early_exit=...)` and `while_loop(max_steps=...)`

### PR DESCRIPTION
**Pull request by patrick-kidger**
TL;DR: bounded reverse-mode differentiable while loops.

Wanting to have an "early exit scan" or a "bounded while loop" is a request that has come up a lot:
https://github.com/google/jax/issues/2139
https://github.com/google/jax/issues/2469
https://github.com/google/jax/issues/5642
https://github.com/google/jax/discussions/8375
https://github.com/google/jax/issues/9960
etc.

In particular for my use-cases, this operation is the one needed inside a differential equation solver. Up until now Diffrax had had a [pretty complicated/inefficient/hacky workaround](https://github.com/google/jax/issues/2139#issuecomment-1039293633).

So anyway! I'm happy to say that this has finally become a big enough problem that I sat down to implement this properly. Building in support for this in Diffrax, I've even seen some nice compile-time improvements: the best-so-far, proportionally, is a reduction from 28 seconds to 5 seconds, just by switching out the bounded while loop implementation.

WDYT?

### CODE REVIEW

File: jax/_src/lax/control_flow/loops.py [hunk 1]
```diff
@@ -155,10 +158,6 @@ def scan(f, init, xs, length=None):
   dtype (or a nested tuple/list/dict container data structure with a fixed
   structure and arrays with fixed shape and dtype at the leaves).
 
-  .. note::
-    :py:func:`scan` compiles ``f``, so while it can be combined with
-    :py:func:`jit`, it's usually unnecessary.
-
```
Comment (patrick-kidger) [inline jax/_src/lax/control_flow/loops.py:161]: Drive-by: I removed these whilst I was editing the docstrings, as right now JIT'd code is the only fast path, so I think this advice is contrary to good practice.

File: jax/_src/lax/control_flow/loops.py [hunk 2]
```diff
@@ -219,18 +223,40 @@ def scan(f, init, xs, length=None):
...
+      pass
+
+  def aug_f(carry, x):
+    out = f(carry, x)
+    if not hasattr(out, "__len__") or len(out) != 2:
+      msg = "scan body output must be a pair, got {}."
+      raise TypeError(msg.format(out))
+    carry, y = out
+    done = early_exit(carry)
+    return done, carry, y
```
Comment (patrick-kidger) [inline jax/_src/lax/control_flow/loops.py:259]: As you can see, the approach I took was to add the conditional as another output of the body, rather than as a separate jaxpr.

Upside: this really helped to minimise the amount of code changes needed elsewhere. No keeping track of separate constants for both `f` and `early_exit`, no extra fixed-point resolution needed, etc.

Downside: `early_exit` is traced/compiled twice; once at initialisation and once in the body.

Conditionals are generally small and cheap, so I felt this was a worthy trade-off here.

File: jax/_src/lax/control_flow/loops.py [hunk 3]
```diff
@@ -263,106 +286,197 @@ def _create_jaxpr(init):
...
   disallowed_effects = jaxpr.effects - allowed_effects
   if disallowed_effects:
     raise NotImplementedError(
         f'Effects not supported in `scan`: {disallowed_effects}')
 
-  out = scan_p.bind(*consts, *in_flat,
+  # use default precision with no weak types
+  start = jax.numpy.array([0])[0]
+  stop = jax.numpy.array([length])[0]
+  done = jax.numpy.array(early_exit(init), dtype=np.bool_)
```
Comment (patrick-kidger) [inline jax/_src/lax/control_flow/loops.py:299]: I suspect there's a better way to accomplish this but I don't know what it is.

Comment (patrick-kidger) [reply jax/_src/lax/control_flow/loops.py:299]: Besides introducing the extra `done` variable, we also introduce `start` and `stop`. The iteration happens over 'xs[start:stop]`. (In particular inclusive of `start` but exclusive of `stop`.) This is needed when backpropagating through an early-exit'd scan: we would like to start from partway through the backward iterations.

File: jax/_src/lax/control_flow/loops.py [hunk 4]
```diff
@@ -263,106 +286,197 @@ def _create_jaxpr(init):
...
+                    linear=(False,) * (3 + len(consts) + len(in_flat)),
                     unroll=unroll)
-  return tree_unflatten(out_tree, out)
-
-def _scan_impl_unrolled(*args, reverse, length, num_consts, num_carry, linear,
-                        f_impl, x_avals, y_avals):
-  consts, init, xs = split_list(args, [num_consts, num_carry])
+  out = out[2:]  # trim off start and stop
+  _, carry, ys = tree_unflatten(out_tree, out)  # trim off done
+  return carry, ys
```
Comment (patrick-kidger) [inline jax/_src/lax/control_flow/loops.py:307]: The primitive returns `start`, `stop`, and `done` for use as residuals in the rules below. (And so they need to be trimmed off here before returning to the user.)

File: jax/_src/lax/control_flow/loops.py [hunk 5]
```diff
@@ -797,58 +1111,84 @@ def _scan_dce_rule(used_outputs: List[bool], eqn: core.JaxprEqn
       used_carry_out = _map(operator.or_, used_carry_out, used_carry_in)
   else:
     assert False, "Fixpoint not reached"
-  if config.jax_enable_checks: core.check_jaxpr(jaxpr.jaxpr)
+  if config.jax_enable_checks: core.check_jaxpr(jaxpr_dce)
```
Comment (patrick-kidger) [inline jax/_src/lax/control_flow/loops.py:1114]: Drive-by: I think this was just a bug in the original code.

File: jax/_src/lax/control_flow/loops.py [hunk 6]
```diff
@@ -797,58 +1111,84 @@ def _scan_dce_rule(used_outputs: List[bool], eqn: core.JaxprEqn
...
+  [start_uk, stop_uk, done_uk], const_uk, carry_uk, xs_uk = split_list(unks_in, [3, num_consts, num_carry])
+  if done_uk:
+    # If conditional is unknown, just do the default processing.
+    eqn_known = None
+    eqn_staged = eqn
+    unks_out = [True] * len(eqn.outvars)
+    inst_out = [True] * len(eqn.outvars)
+    new_vars = [x for inst, x in zip(inst_in, eqn.invars)
+                if type(x) is core.Var and not inst]
+    return eqn_known, eqn_staged, unks_out, inst_out, new_vars
```
Comment (patrick-kidger) [inline jax/_src/lax/control_flow/loops.py:1146]: I don't think there's a hook for this default processing, so I essentially just ended up copying what `pe.partial_eval_jaxpr_custom` does by default.

File: jax/core.py [hunk 7]
```diff
@@ -2464,9 +2464,9 @@ class JaxprTypeError(TypeError): pass
 
 custom_typechecks: Dict[Primitive, Callable] = {}
 
-def _check_closed_call(*in_atoms, call_jaxpr):
+def _check_closed_call(*in_atoms, call_jaxpr, name=None):
```
Comment (patrick-kidger) [inline jax/core.py:2467]: Drive-by bugfix, that meant I was otherwise getting a crash.

File: tests/lax_control_flow_test.py [hunk 8]
```diff
@@ -2248,6 +2256,119 @@ def f(c, a):
         len(str(jax.xla_computation(scan)(c, xs).as_hlo_text())),
         len(str(jax.xla_computation(scan_unrolled)(c, xs).as_hlo_text())))
 
+  @parameterized.named_parameters(
+      {"testcase_name": f"_{scan_name}",
+       "scan": scan_impl}
+      for scan_impl, scan_name in SCAN_IMPLS + [(scan_reference, "reference")])
```
Comment (patrick-kidger) [inline tests/lax_control_flow_test.py:2262]: The reference implementation is included to test the test.

**Comment by patrick-kidger** [comment]
Hmmm, looks like there's some differences between me running pytest locally and the way it's ran here. I'll have to start digging through these test errors and figure out what's going on.

**Comment by jakevdp** [comment]
It looks like an X64 test failing; you could try `JAX_ENABLE_X64=1 pytest -n auto tests`

**Comment by mattjj** [comment]
I'd be happy to start taking a look, but I'm on paternity leave until end of November!

**Comment by patrick-kidger** [comment]
Reporting in to say that I've now benchmarked this on our production code, and that this has improved our runtime from 10 seconds-per-iteration down to 6 seconds-per-iteration. (!)

**Comment by carlosgmartin** [comment]
@patrick-kidger Could an early exit be added to [`fori_loop`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.fori_loop.html) as well (#8375)? This would be useful for models that run for a variable number of iterations, e.g.
1. [Can You Learn an Algorithm? Generalizing from Easy to Hard Problems with Recurrent Networks](https://arxiv.org/abs/2106.04537)
2. [PonderNet: Learning to Ponder](https://arxiv.org/abs/2107.05407)
3. [End-to-end Algorithm Synthesis with Recurrent Networks: Logical Extrapolation Without Overthinking](https://arxiv.org/abs/2202.05826)

**Comment by epignatelli** [comment]
I understand this is old and probably superseded. Has this been implemented?
If not, is this PR going anywhere?

**Comment by patrick-kidger** [comment]
Unfortunately I believe this PR is dead. In the end the decision was that this was a change to a central part of the JAX codebase, and that was probably too risky to merge.

**Fortunately**, there is an alternative! Try `equinox.internal.while_loop(..., kind="checkpointed")`. This is a backpropagatable while loop using recursive checkpointing.

**Comment by epignatelli** [comment]
That's a pity, there could be uses cases of interest. Jitting recursive sampling with a termination criterion is the first that comes to mind (LLMs?).

I'll have a look at the `equinox.internal.while_loop(..., kind="checkpointed")`, thanks @patrick-kidger !

**Comment by carlosgmartin** [comment]
I'd love to see this functionality in JAX, as well.
